### DISCUSSION TITLE
Discrete bounds for integer variables

### DISCUSSION THREAD

**Discussion by juan11iguel** [esa/pygmo2#175] [Ideas]
As of now, the bounds in the `Problem` interface can only be set as box-bounds:

> The `get_bounds()` method of the UDP must return the box-bounds as a tuple of 2 elements, the lower bounds vector and the upper bounds vector

This is fine for real variables, but for integer ones, where the value may represent for example a finite machine state, it would be very beneficial having the ability to restrict the possible states to a set of **discrete** values.

Pseudo-code of how this interface could look:
```python
...
def get_bounds(self) -> tuple[ list[float|list[int]], list[float|None] ]
    # Variables 0,1,2 are real
    lower_bounds = [0, 0, -np.inf]
    upper_bounds = [10, 100, np.inf]
    # Variable 4 is an integer
    lower_bounds.append( [0,3,4] )
    upper_bounds.append( None )

    return lower_bounds, upper_bounds
```

With the current implementation, one would have to define [0,4] as the bounds for the discrete variable, adding complexity to the solution space (1,2 would be explored) that could have otherwise been avoided by setting appropriate bounds.

Does what I propose make any sense? I understand that this would imply a modification on how the initial values are generated, which for these integer variables with bounds would need to pick one element from the set of possible values at random. 

Best regards!

**Comment by darioizzo** [comment]
Of course, what you want can be obtained mapping the values [0,1,2] to your finite set [0,3,4]. The bounds would then be [0,2] and the mapping used inside the fitness function. Makes sense?

**Comment by juan11iguel** [comment]
Yes!
... or at least I think so 😅. Here is some pseudo-code of a possible implementation:

```python
class Problem:
    
    logical_dec_vars_mapping: dict[str, list[int]] = {} # Probably throws error due to mutable default value
    
    def get_bounds(self) -> tuple[ list[float|int], list[float|int] ]:
        # Variables 0,1,2 are real
        lower_bounds = [0, 0, -np.inf]
        upper_bounds = [10, 100, np.inf]
        
        # Variable 3 is an integer with 5 possible states
        if some_condition:
            # allowed states [0,3,4] -> bounds: [0, 2]
            lower_bounds.append(0)
            upper_bounds.append(2)
            # Setup mapping
            self.logical_dec_vars_mapping["var_3"] = [0,3,4]
        
        else: # Default case
            # allowed states [0,1,2,3,4] -> bounds: [0, 4]
            lower_bounds.append(0)
            upper_bounds.append(4)
            # Setup default mapping
            self.logical_dec_vars_mapping["var_3"] = [0,1,2,3,4]

        return lower_bounds, upper_bounds

    def fitness(self, x: np.ndarray[float | int]):
        
        # Map logical decision variables values to their actual values
        x[3] = self.logical_dec_vars_mapping["var_3"][x[3]]
        
        ...
```

Thanks!
